wallet/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+        let mut is_existing_payment = false;
+        let mut signer_ids = vec![&env];
+        if let Some(maybe_previous_signers) = env.data().get(&DataKey::PaySigners(payment_id)) {
+            is_existing_payment = true;
+            // If there were previous signers for this payment id, we need to check that
+            // the payment still hasn't been executed (it should be removed on execution)
+            // and that it matches the payment signed by the new signers.
+            let stored_payment: Payment = env
+                .data()
+                .get_unchecked(&DataKey::Payment(payment_id))
+                .unwrap();
+            if stored_payment != payment {
+                panic!("stored payment doesn't match new payment with same id");
+            }
+            let previous_signers: WeightedSigners = maybe_previous_signers.unwrap();
+            signer_ids = previous_signers.signers;
+            // Check that no new signers have already signed this payment and
+            // panic if that's not the case.
+            // This is only one option of how to handle this; an alternative approach
+            // сould be to only account for weight of the new signers, but that's likely
+            // more error-prone (there shouldn't be a reason for an admin to
+            // resubmit the signature).
+            for maybe_signature in signatures.iter() {
+                let id = maybe_signature.unwrap().identifier(&env);
+                if signer_ids.contains(&id) {
+                    panic!("one of the signers has already signed this payment");
+                }
+            }
+            weight_sum += previous_signers.weight;
+        }
+
+        for maybe_signature in signatures.iter() {
+            signer_ids.push_back(maybe_signature.unwrap().identifier(&env));
+        }
+        // Update signer data. This also serves as a protection from
+        // re-executing the payment with the same id (a separate entry could
+        // serve this purpose as well).
+        env.data().set(
+            DataKey::PaySigners(payment_id),
+            WeightedSigners {
+                signers: signer_ids,
+                weight: weight_sum,
+            },
+        );
+
+        let threshold = read_threshold(&env);
+        // When there is enough signature weight to authorize this payment
+        // execute the payment immediately.
+        if weight_sum >= threshold {
+            execute_payment(&env, payment);
+            // Remove the payment to mark it executed (signers are still there).
+            env.data().remove(&DataKey::Payment(payment_id));
+            return true;
+        }
+        if !is_existing_payment {
+            env.data().set(DataKey::Payment(payment_id), payment);
+        }
+
+        false
+    }
+}
+
+fn check_initialization_params(env: &Env, admins: &Vec<Admin>, threshold: u32) {
+    if threshold == 0 {
+        panic!("threshold has to be non-zero");
```

Comment:
Given the amount of input validation this contract would benefit from using `panic_error!` I think, so that if there are panics in the final compiled, there are contract error codes rather than the trap error code. See https://soroban-dev.stellar.org/docs/examples/errors#panicking-with-an-error.

Reply:
I agree that errors would be nice. Moreover, I would probably rewrite this to return `Result` instead of panicking and use `try_` functions in tests as `should_panic` is pretty annoying. The only thing I'm not sure about is granularity: replacing every panic with a unique error is a bit annoying (and also inconvenient for a potential consumer of these errors), while reducing granularity would lose some information/test coverage (now I can precisely set expectations on why should the contract fail). So ideally I'd want to both return an error and have a string explanation (also as a debug event?) for debugging/testing.

Reply:
I'd lean into granularity.

You can still use the `try` functions if you use the `panic_error!` macro. This might be a good balance if the intent is mostly for testing and diagnostics, and less for developers to handle them explicitly. The only difference between returning Result or not is that the Error gets encoded into the contract spec. So if you want people to handle errors, getting it into the contract spec helps to document it.

> So ideally I'd want to both return an error and have a string explanation (also as a debug event?) for debugging/testing.

We might want to explore modifying `panic_error!` so that it can optionally take a string as well, and it can log a debug event using `log!` alongside storing the error.

Reply:
> I'd lean into granularity.

True, I guess unlike with host more granular errors are more acceptable for contracts (given that they're immutable). 

> We might want to explore modifying panic_error! so that it can optionally take a string as well, and it can log a debug event using log! alongside storing the error.

That's what I've been thinking about, however currently the debug events from `log` don't seem to appear in the test logs, so we need to think how to make this useful. For now I'll just use granular errors.

Reply:
You can make the logs appear in test logs by adding this line to the end of your test:
```rust
env.logger().print()
```

I'd love to find a way to hook this in automatically if you have any idea.

Reply:
> The only difference between returning Result or not is that the Error gets encoded into the contract spec.

It seems like another different thing is that `panic_error` still panics, which makes tests harder (and IIRC we can't `catch_unwind` that). So for the sake of testing `Result` seems preferential, as `should_panic` is inconvenient when there are a lot of failure scenarios to cover.

Reply:
> `panic_error` still panics, which makes tests harder

Agreed, and there's an issue open to address this: 
- https://github.com/stellar/rs-soroban-sdk/issues/642
- https://github.com/stellar/rs-soroban-env/issues/430

Reply:
Fwiw I think returning Results from these functions would also be very reasonable. I'm really curious if you add that how much the contract size changes. I think we've been worried return paths could be expensive, but it's all assumptions at this point, and you might prove that wrong, which imo would be awesome. So awesome we could maybe even axe panic_error!.

Reply:
In https://github.com/stellar/soroban-examples/pull/150 I'm using results everywhere besides `ConversionError` unwraps. I think using result doesn't make things much more complex/verbose, It definitely is beneficial for testing and the code is IMHO more idiomatic Rust.